lock-api/src/main/java/com/palantir/lock/client/timestampleases/TimestampLeaseAcquirerImpl.java

```diff
@@ -16,28 +16,203 @@
+import com.palantir.common.exception.AtlasDbDependencyException;
+import com.palantir.common.streams.KeyedStream;
+import com.palantir.lock.ConjureTimestampRangeTimestampSupplier;
+import com.palantir.lock.LimitingLongSupplier;
+import com.palantir.lock.client.LeasedLockToken;
+import com.palantir.lock.v2.TimestampLeaseResult;
 import com.palantir.lock.v2.TimestampLeaseResults;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeRuntimeException;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.tritium.ids.UniqueIds;
+import java.util.List;
 import java.util.Map;
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.function.LongSupplier;
+import java.util.function.Supplier;
+import org.immutables.value.Value;
+import org.jetbrains.annotations.VisibleForTesting;
 
 final class TimestampLeaseAcquirerImpl implements TimestampLeaseAcquirer {
+    private static final SafeLogger log = SafeLoggerFactory.get(TimestampLeaseAcquirerImpl.class);
+
     private final NamespacedTimestampLeaseService delegate;
-    private final TimeLockUnlocker unlocker;
+    private final Unlocker unlocker;
+    private final Supplier<UUID> uuidSupplier;
+
+    private final Retryer<RequestAndResponse> retryer = RetryerBuilder.<RequestAndResponse>newBuilder()
+            .retryIfResult(requestAndResponse -> !wasRequestFullyFulfilled(requestAndResponse))
+            .retryIfException()
+            .withStopStrategy(StopStrategies.stopAfterAttempt(3))
+            .build();
 
-    public TimestampLeaseAcquirerImpl(NamespacedTimestampLeaseService delegate, TimeLockUnlocker unlocker) {
+    @VisibleForTesting
+    TimestampLeaseAcquirerImpl(
+            NamespacedTimestampLeaseService delegate, Unlocker unlocker, Supplier<UUID> uuidSupplier) {
         this.delegate = delegate;
         this.unlocker = unlocker;
+        this.uuidSupplier = uuidSupplier;
+    }
+
+    public TimestampLeaseAcquirerImpl(NamespacedTimestampLeaseService delegate, Unlocker unlocker) {
+        this(delegate, unlocker, UniqueIds::pseudoRandomUuidV4);
     }
 
     @Override
     public TimestampLeaseResults acquireNamedTimestampLeases(Map<TimestampLeaseName, Integer> requests) {
-        // TODO(aalouane): implement
-        throw new UnsupportedOperationException("Not implemented yet");
+        TimestampLeaseResponses response = acquireLeasesWithRetry(requests).response();
+        try {
+            return TimestampLeaseResults.builder()
+                    .lock(createLeasedLockToken(response))
+                    .results(createTimestampLeaseResult(requests, response.getTimestampLeaseResponses()))
+                    .build();
+        } catch (RuntimeException | Error e) {
+            log.error("Unexpected exception while creating client results", e);
+            asyncUnlock(response);
```

Comment:
I would very strongly prefer sync unlock here: it's easey to get into a very nasty retry storm

Reply:
Sorry, this is supposed to be sync and it is sync in the next PR. I just did not update the unlock